### PR TITLE
chore: Add GH action for regenerating code from API spec

### LINKED ISSUE #54: Automated Spec syncing
We should simplify the approach taken to updating the API specification (as seen in [metal-go](https://github.com/equinix-labs/metal-go/actions/workflows/sync.yml), and [metal-java](https://github.com/equinix-labs/metal-java/actions/workflows/sync.yml)).

There should be a GitHub action to trigger which will fetch the latest spec and update the repo in a PR.

### PR DESCRIPTION
**Pull request by t0mk**
This PR adds workflow for automatically regeneratign the SDK based on latest API spec, inspired by similar action from metal-go:

https://github.com/equinix-labs/metal-go/blob/main/.github/workflows/sync.yml

Fixes #54

### CODE REVIEW

File: .github/workflows/sync.yml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        # https://api.github.com/users/github-actions[bot]
+        git config user.name 'github-actions[bot]'
+        git config user.email '41898282+github-actions[bot]@users.noreply.github.com'
+    - name: Fetch latest spec to oas.fetched/
+      id: fetch
+      run: |
+        make fetch
+        git add oas3.fetched
+        echo `git commit -m 'sync: fetch ${{ steps.date.outputs.date }} spec and apply patches'`
+    - name: Build single-file spec in oas3.stitched/metal_openapi.yaml
```
Comment (ctreatma) [inline .github/workflows/sync.yml:30]: We removed this step in `metal-go` (and it turned out that combining the spec might have introduced a bug in how `include` and `exclude` were handled).  Removing it now might introduce some minor code changes & make this PR bigger, so if you'd prefer to keep it as-is and remove the stitch step later, that's fine.

File: .github/workflows/sync.yml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      if: ${{ always() && steps.fetch.conclusion == 'success' }}
+      run: |
+        make stitch-spec
+        git add oas3.stitched
+        echo `git commit -m 'sync: stitch spec with ${{ steps.date.outputs.date }} spec'`
+    - name: Patch spec and build ./metal_openapi.fixed.yml
+      id: patch
+      if: ${{ always() && steps.stitch.conclusion == 'success' }}
+      run: |
+        make patch-spec
```
Comment (ctreatma) [inline .github/workflows/sync.yml:41]: Ah, right...I forgot that metal-python patches the stitched spec, not the split spec; in that case, it'll be a much bigger change to abandon the stitch step here.

File: .github/workflows/sync.yml [hunk 3]
```diff
@@ -0,0 +1,92 @@
+name: Sync
+
+on:
+  workflow_dispatch:
+
+jobs:
+  sync:
+    strategy:
+      matrix:
+        go-version: [1.19.x]
```
Comment (ctreatma) [inline .github/workflows/sync.yml:10]: I don't think you need this Go version, but there does need to be a Python version & setup step (unless that's baked in somewhere else?).

**Comment by t0mk** [comment]
I'm testing locally, this is still Work In Progress

**Comment by t0mk** [comment]
I need https://github.com/equinix-labs/metal-python/pull/37 merged before I can test this.

**Review by ctreatma — APPROVED** [review]
The test sync PR I saw for this workflow looked good

**Comment by t0mk** [comment]
I managed to run the action locally with my personal token, it succeeded and created the PR:
https://github.com/equinix-labs/metal-python/pull/41

.. so this is ready for review.

As for the account that creates the PRs, I used the same as in metal-go"

`author: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>`

@ctreatma Do we need the github token secret configured here? Or will it assume credentials of the executor? I guess you know because you did it in metal-go.

**Comment by ctreatma** [comment]
The sync action uses [the built-in, short-lived token provided by GitHub Actions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#about-the-github_token-secret), so no additional secret is needed to run the job.

**Comment by ctreatma** [comment]
I made a minor update to the fetch task definition and ran the sync job on my fork.  Sample PR generated by the updated sync job is here: https://github.com/ctreatma/metal-python/pull/1

**Review by ctreatma — APPROVED** [review]

**Pull request merged**